File: android/app/src/main/java/com/winchester/marquee/MarqueeWidget.kt
package com.winchester.marquee

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.view.View
import android.widget.RemoteViews
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.time.Duration
import java.time.OffsetDateTime
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlin.concurrent.thread

/**
 * Home-screen widget for the Winchester marquee.
 *
 * Reads the same cached snapshot the web board reads and renders the verdict
 * already computed by the fetcher. It never decides anything itself, so the
 * widget cannot disagree with the grid or the board about whether a title is
 * dimmed.
 *
 * Replaces the Scriptable script in widget/, which is iOS-only.
 *
 * NOT COMPILED. This was written without an Android SDK available, so it has
 * never been built or run. Treat it as reviewed, not working: open the
 * android/ directory in Android Studio, let it sync, and expect to fix
 * something on the first build.
 */
class MarqueeWidget : AppWidgetProvider() {

    companion object {
        /**
         * Where web/ is being served from.
         *
         * Defaults to the phone itself, which is the common setup: Termux
         * runs the fetcher and a loopback HTTP server, so the widget, the
         * page and the box are all one device and nothing touches the
         * network. Point it at a hostname instead if you run the fetcher on
         * a Pi — and add that hostname to network_security_config.xml, or
         * Android will refuse the cleartext request.
         */
        const val BASE_URL = "http://127.0.0.1:8080"

        private const val DATA_PATH = "/data/marquee.json"
        private const val PAGE_PATH = "/board.html"
        private const val PREFS = "marquee"
        private const val KEY_LAST_GOOD = "last_good_json"
        private const val ACTION_REFRESH = "com.winchester.marquee.REFRESH"

        /** Row slots declared in widget_marquee.xml. */
        private val ROW_IDS = intArrayOf(
            R.id.row0, R.id.row1, R.id.row2, R.id.row3, R.id.row4, R.id.row5
        )
        private val TITLE_IDS = intArrayOf(
            R.id.title0, R.id.title1, R.id.title2,
            R.id.title3, R.id.title4, R.id.title5
        )
        private val TIME_IDS = intArrayOf(
            R.id.time0, R.id.time1, R.id.time2,
            R.id.time3, R.id.time4, R.id.time5
        )
    }

    override fun onUpdate(
        context: Context,
        manager: AppWidgetManager,
        widgetIds: IntArray
    ) {
        refresh(context, manager, widgetIds)
    }

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)
        if (intent.action == ACTION_REFRESH) {
            val manager = AppWidgetManager.getInstance(context)
            val ids = manager.getAppWidgetIds(
                android.content.ComponentName(context, MarqueeWidget::class.java)
            )
            refresh(context, manager, ids)
        }
    }

    private fun refresh(
        context: Context,
        manager: AppWidgetManager,
        widgetIds: IntArray
    ) {
        if (widgetIds.isEmpty()) return
        // Widget callbacks run on the main thread and the host kills slow
        // ones, so the fetch goes to a background thread and the views are
        // pushed when it returns.
        thread {
            val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            val result = fetch()
            var body = result.body
            var stale = false

            if (body == null) {
                // Stale beats blank, the same rule the cron and the page use.
                body = prefs.getString(KEY_LAST_GOOD, null)
                stale = true
            } else {
                prefs.edit().putString(KEY_LAST_GOOD, body).apply()
            }

            val views = render(context, body, stale, result.problem)
            for (id in widgetIds) manager.updateAppWidget(id, views)
        }
    }

    /** What went wrong, in the words the person holding the phone needs. */
    private class Fetched(val body: String?, val problem: String?)

    private fun fetch(): Fetched = try {
        val connection = (URL(BASE_URL + DATA_PATH).openConnection() as HttpURLConnection)
        connection.connectTimeout = 8000
        connection.readTimeout = 8000
        connection.useCaches = false
        val code = connection.responseCode
        when {
            code == 404 -> Fetched(null, "Server is up but has no snapshot yet. Run refresh.py.")
            code >= 400 -> Fetched(null, "Server answered HTTP $code.")
            else -> Fetched(connection.inputStream.bufferedReader().use { it.readText() }, null)
        }
    } catch (e: java.net.ConnectException) {
        // Overwhelmingly the common case: nothing is listening, because the
        // Termux server is not running. Saying so beats "cannot reach".
        Fetched(null, "Nothing is listening on $BASE_URL — is the server running?")
    } catch (e: java.net.SocketTimeoutException) {
        Fetched(null, "$BASE_URL timed out.")
    } catch (e: java.io.IOException) {
        // Cleartext blocked by the network security config lands here, and it
        // is invisible otherwise — the host has to be listed in
        // network_security_config.xml or Android refuses plain HTTP outright.
        Fetched(null, "Cannot read $BASE_URL: ${e.javaClass.simpleName}. " +
            "If you changed the host, add it to network_security_config.xml.")
    } catch (e: Exception) {
        Fetched(null, "${e.javaClass.simpleName}: ${e.message ?: "no detail"}")
    }

    private fun render(
        context: Context,
        body: String?,
        stale: Boolean,
        problem: String?
    ): RemoteViews {
        val views = RemoteViews(context.packageName, R.layout.widget_marquee)

        // Tapping opens the board, which is where a verdict can explain
        // itself. A widget has no room for the reason text.
        val open = PendingIntent.getActivity(
            context, 0,
            Intent(Intent.ACTION_VIEW, Uri.parse(BASE_URL + PAGE_PATH)),
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )
        views.setOnClickPendingIntent(R.id.widget_root, open)

        for (id in ROW_IDS) views.setViewVisibility(id, View.GONE)

        if (body == null) {
            views.setTextViewText(R.id.stamp, "NO SIGNAL")
            views.setViewVisibility(R.id.empty, View.VISIBLE)
            // Never just "cannot reach": with no cache to fall back on this
            // text is the only thing to debug from, and the widget cannot
            // show a stack trace.
            views.setTextViewText(R.id.empty, problem ?: "No cached snapshot yet.")
            return views
        }

        val rows: List<Row>
        val snapshot: JSONObject
        try {
            snapshot = JSONObject(body)
            rows = upcoming(snapshot)
        } catch (e: Exception) {
            views.setTextViewText(R.id.stamp, "BAD DATA")
            views.setViewVisibility(R.id.empty, View.VISIBLE)
            views.setTextViewText(R.id.empty, "Snapshot could not be read.")
            return views
        }

        val fetchedAt = snapshot.optString("fetched_at", "")
        val age = relativeAge(fetchedAt)
        val isStale = stale || snapshot.optBoolean("stale", false)
        views.setTextViewText(R.id.stamp, if (isStale) "STALE $age" else "UPDATED $age")
        views.setTextColor(
            R.id.stamp,
            context.getColor(if (isStale) R.color.alert else R.color.ink_soft)
        )

        if (rows.isEmpty()) {
            views.setViewVisibility(R.id.empty, View.VISIBLE)
            views.setTextViewText(R.id.empty, "Nothing left on the board.")
            return views
        }
        views.setViewVisibility(R.id.empty, View.GONE)

        val ink = context.getColor(R.color.flap_ink)
        val inkDim = context.getColor(R.color.flap_ink_dim)

        for ((slot, row) in rows.take(ROW_IDS.size).withIndex()) {
            views.setViewVisibility(ROW_IDS[slot], View.VISIBLE)
            // A trailing ? means the rating reason could not be read: unknown,
            // not clean. Never silently treated as fine.
            val mark = if (row.unknown) "  ?" else ""
            views.setTextViewText(TITLE_IDS[slot], row.name.uppercase(Locale.US) + mark)
            views.setTextViewText(TIME_IDS[slot], "${row.day} ${row.time}")

            // Dimming fades the letters, it does not darken the sign.
            val colour = if (row.flagged) inkDim else ink
            views.setTextColor(TITLE_IDS[slot], colour)
            views.setTextColor(TIME_IDS[slot], colour)
        }
        return views
    }

    private data class Row(
        val name: String,
        val time: String,
        val day: String,
        val flagged: Boolean,
        val unknown: Boolean,
        val at: OffsetDateTime
    )

    /**
     * One entry per title, using its soonest showing that has not started.
     * Once the day's last screening has gone the row rolls onto tomorrow by
     * itself, so the widget never advertises a screening already missed.
     */
    private fun upcoming(snapshot: JSONObject): List<Row> {
        val now = OffsetDateTime.now()
        val titles = snapshot.optJSONArray("titles") ?: return emptyList()
        val rows = ArrayList<Row>()

        for (i in 0 until titles.length()) {
            val title = titles.optJSONObject(i) ?: continue
            val showings = title.optJSONArray("showings") ?: continue

            var soonest: OffsetDateTime? = null
            for (j in 0 until showings.length()) {
                val raw = showings.optJSONObject(j)?.optString("showtime") ?: continue
                val at = try {
                    OffsetDateTime.parse(raw)
                } catch (e: Exception) {
                    continue
                }
                if (at.isBefore(now)) continue
                if (soonest == null || at.isBefore(soonest)) soonest = at
            }
            val at = soonest ?: continue

            rows.add(
                Row(
                    // display_name drops Alamo's booking decoration; a widget
                    // row has no characters to spare on a strand name.
                    name = title.optString(
                        "display_name",
                        title.optString("name", title.optString("slug", "?"))
                    ),
                    time = at.format(DateTimeFormatter.ofPattern("h:mma", Locale.US))
                        .replace("AM", "A").replace("PM", "P"),
                    day = dayLabel(at),
                    flagged = title.optBoolean("flagged", false),
                    unknown = !title.optBoolean("reason_parsed", false),
                    at = at
                )
            )
        }
        rows.sortBy { it.at }
        return rows
    }

    private fun dayLabel(at: OffsetDateTime): String {
        val today = ZonedDateTime.now().toLocalDate()
        val date = at.toLocalDate()
        return when (date) {
            today -> "TODAY"
            today.plusDays(1) -> "TMRW"
            else -> date.dayOfWeek
                .getDisplayName(java.time.format.TextStyle.SHORT, Locale.US)
                .uppercase(Locale.US)
        }
    }

    private fun relativeAge(iso: String): String {
        if (iso.isEmpty()) return ""
        return try {
            val minutes = Duration.between(OffsetDateTime.parse(iso), OffsetDateTime.now())
                .toMinutes()
                .coerceAtLeast(0)
            when {
                minutes < 60 -> "${minutes}M"
                minutes < 60 * 24 -> "${minutes / 60}H"
                else -> "${minutes / (60 * 24)}D"
            }
        } catch (e: Exception) {
            ""
        }
    }
}

File: widget/marquee-widget.js
// Winchester Marquee — Scriptable home-screen widget (iOS).
//
// Install: put this file in Scriptable's folder (Files > Scriptable), add a
// Scriptable widget to the home screen, and set its script to this one. Set
// MARQUEE_URL below to the box serving web/.
//
// The widget is the glance. It cannot host the drill-in panel — iOS widgets
// have no in-widget interaction beyond a tap target — so tapping anywhere
// opens the companion page, which carries the full grid and the "why".
//
// It renders the verdict the fetcher already computed and never decides
// anything itself, so the widget and the page can never disagree about
// whether a title is dimmed.
//
// An Android equivalent reads this identical JSON; nothing here is
// iOS-specific except the Scriptable API calls.

const MARQUEE_URL = 'http://winchester.local:8080';

const DATA_URL = `${MARQUEE_URL}/data/marquee.json`;
const PAGE_URL = `${MARQUEE_URL}/index.html`;
const CACHE_FILE = 'winchester-marquee.json';

const BG = new Color('#0B0B0E');
const TEXT = new Color('#F2F2F0');
const DIM_TEXT = new Color('#F2F2F0', 0.34); // the weighting, on a widget
const MUTED = new Color('#9A9AA6');
const ACCENT = new Color('#E8B33C');
const ALERT = new Color('#E5484D');

// ---------- data, with its own fallback cache ----------

function cachePath() {
  const fm = FileManager.local();
  return fm.joinPath(fm.cacheDirectory(), CACHE_FILE);
}

async function loadData() {
  try {
    const req = new Request(DATA_URL);
    req.timeoutInterval = 8;
    const data = await req.loadJSON();
    FileManager.local().writeString(cachePath(), JSON.stringify(data));
    return data;
  } catch (err) {
    // Box asleep or off the network: stale beats blank, same rule as the page.
    const fm = FileManager.local();
    if (fm.fileExists(cachePath())) {
      const cached = JSON.parse(fm.readString(cachePath()));
      cached.stale = true;
      return cached;
    }
    return null;
  }
}

// ---------- time helpers ----------

const dayKey = iso => iso.slice(0, 10);
const todayKey = () => {
  const d = new Date();
  const local = new Date(d.getTime() - d.getTimezoneOffset() * 60000);
  return local.toISOString().slice(0, 10);
};

function timeLabel(iso) {
  const d = new Date(iso);
  let h = d.getHours();
  const m = d.getMinutes().toString().padStart(2, '0');
  const suffix = h >= 12 ? 'p' : 'a';
  h = h % 12 || 12;
  return m === '00' ? `${h}${suffix}` : `${h}:${m}${suffix}`;
}

function relativeAge(iso) {
  const mins = Math.round((Date.now() - new Date(iso).getTime()) / 60000);
  if (mins < 60) return `${Math.max(mins, 0)}m`;
  const hrs = Math.round(mins / 60);
  return hrs < 24 ? `${hrs}h` : `${Math.round(hrs / 24)}d`;
}

/* Upcoming showings, flattened across titles and sorted by time. Past
   showtimes are dropped — a widget showing this afternoon's 2pm at 9pm is
   worse than showing nothing. */
function upcoming(data, limit) {
  const now = Date.now();
  const rows = [];
  for (const t of data.titles) {
    for (const s of t.showings) {
      const when = new Date(s.showtime).getTime();
      if (when < now) continue;
      rows.push({ title: t, showing: s, when });
    }
  }
  rows.sort((a, b) => a.when - b.when);

  // One row per title — the soonest showing — so a four-show title cannot
  // crowd everything else off a small widget.
  const seen = new Set();
  const out = [];
  for (const row of rows) {
    if (seen.has(row.title.slug)) continue;
    seen.add(row.title.slug);
    out.push(row);
    if (out.length >= limit) break;
  }
  return out;
}

// ---------- rendering ----------

function header(w, data) {
  const row = w.addStack();
  row.centerAlignContent();

  const title = row.addText('WINCHESTER');
  title.font = Font.heavySystemFont(11);
  title.textColor = TEXT;

  row.addSpacer();

  const age = data.stale
    ? `stale ${relativeAge(data.fetched_at)}`
    : relativeAge(data.fetched_at);
  const stamp = row.addText(age);
  stamp.font = Font.mediumSystemFont(9);
  stamp.textColor = data.stale ? ACCENT : MUTED;

  w.addSpacer(7);
}

function showingRow(w, { title, showing }, compact) {
  const row = w.addStack();
  row.centerAlignContent();
  row.spacing = 6;

  // Series colour as a leading rule — one-night programming is why this
  // display exists, so it reads before the title does.
  const rule = row.addStack();
  rule.size = new Size(3, compact ? 13 : 15);
  rule.cornerRadius = 1.5;
  rule.backgroundColor = title.series
    ? new Color(title.series.background)
    : new Color('#2C2C36');

  const dimmed = title.flagged;

  const time = row.addText(timeLabel(showing.showtime));
  time.font = Font.semiboldSystemFont(compact ? 10 : 11);
  time.textColor = dimmed ? DIM_TEXT : TEXT;
  time.lineLimit = 1;

  const name = row.addText(title.display_name || title.name);
  name.font = Font.systemFont(compact ? 10 : 11);
  name.textColor = dimmed ? DIM_TEXT : TEXT;
  name.lineLimit = 1;

  row.addSpacer();

  // Unknown is surfaced, never silently treated as clean.
  if (!title.reason_parsed) {
    const q = row.addText('?');
    q.font = Font.boldSystemFont(compact ? 9 : 10);
    q.textColor = ACCENT;
  } else if (dimmed) {
    const dot = row.addText('•');
    dot.font = Font.boldSystemFont(compact ? 11 : 12);
    dot.textColor = new Color(ALERT.hex, 0.75);
  }

  w.addSpacer(compact ? 3 : 4);
}

function footer(w, data, shownCount) {
  const total = data.titles.length;
  if (total <= shownCount) return;
  w.addSpacer(1);
  const more = w.addText(`+${total - shownCount} more`);
  more.font = Font.systemFont(9);
  more.textColor = MUTED;
}

async function build() {
  const w = new ListWidget();
  w.backgroundColor = BG;
  w.setPadding(12, 13, 12, 13);
  w.url = PAGE_URL; // tap -> the page that can actually explain itself

  const data = await loadData();

  if (!data) {
    const t = w.addText('Winchester');
    t.font = Font.heavySystemFont(12);
    t.textColor = TEXT;
    w.addSpacer(4);
    const e = w.addText('No cache yet — is the box reachable?');
    e.font = Font.systemFont(10);
    e.textColor = MUTED;
    return w;
  }

  const family = config.widgetFamily || 'medium';
  const compact = family === 'small';
  const limit = family === 'small' ? 3 : family === 'large' ? 9 : 5;

  header(w, data);

  const rows = upcoming(data, limit);
  if (rows.length === 0) {
    const none = w.addText('Nothing left today.');
    none.font = Font.systemFont(11);
    none.textColor = MUTED;
  } else {
    for (const row of rows) showingRow(w, row, compact);
    footer(w, data, rows.length);
  }

  w.addSpacer();
  return w;
}

const widget = await build();
if (config.runsInWidget) {
  Script.setWidget(widget);
} else {
  await widget.presentMedium();
}
Script.complete();
